ratpack-core/src/main/java/ratpack/health/HealthCheckHandler.java

```diff
@@ -0,0 +1,435 @@
+   * <ul>
+   *   <li><strong>0</strong> - infinite potential parallelism - up to the event loop number of threads</li>
+   *   <li><strong>1</strong> - serial, promises run one by one in sequence</li>
+   *   <li><strong>2</strong> - max 2 at a time promises run in parallel</li>
+   *   <li><strong>3</strong> - max 3 at a time promises run in parallel</li>
+   *   <li><strong>n</strong> - max n at a time promises run in parallel</li>
+   * </ul>
+   */
+  protected HealthCheckHandler(String healthCheckName, int concurrencyLevel) {
+    this.name = healthCheckName;
+    this.concurrencyLevel = concurrencyLevel;
+  }
+
+  /**
+   * Get the level of parallelism in non-blocking health check execution.
+   *
+   * @return concurrency level as defined in {@link ratpack.health.HealthCheckHandler#HealthCheckHandler} constructor. Parameter {@code concurrencyLevel}.
+   */
+  public int getConcurrencyLevel() {
+    return concurrencyLevel;
+  }
+
+  /**
+   * Get assigned health check name. If defined health check of this name is executed.
+   *
+   * @return health check name to be executed
+   */
+  public String getName() {
+    return name;
+  }
+
+  /**
+   * Run individual health check with the given name or all health checks.
+   * The {@code concurrencyLevel} determines parallelism in health check execution.
+   *
+   * @param context request context and exec control as well
+   * @throws Exception if anything goes wrong, exception will be implicitly passed to the context's {@link Context#error(Throwable)} method
+   */
+  @Override
+  public void handle(Context context) throws Exception {
+    if (!name.equals(DEFAULT_NAME_TOKEN)) {
+      handleByName(context, name);
+    } else {
+      handleAll(context);
+    }
+  }
+
+  /**
+   * Run health check with the given name. If it is not registered HTTP 404 is send to the client
+   * All exceptions from health check are wrapped in unhealthy result with error=Exception.
+   * @param context request context
+   * @param name health check name
+   * @throws Exception
+   */
+  private void handleByName(Context context, String name) throws Exception {
+    if (name == null || "".equals(name) || DEFAULT_NAME_TOKEN.equals(name)) {
+      context.clientError(404);
+      return;
+    }
+    SortedMap<String, HealthCheck.Result> hcheckResults = new ConcurrentSkipListMap<>();
+    Optional<HealthCheck> hcheck = context.first(TypeToken.of(HealthCheck.class), hc -> hc.getName().equals(name));
+    if (!hcheck.isPresent()) {
+      context.clientError(404);
+      return;
+    }
+    try {
+      Promise<HealthCheck.Result> promise = hcheck.get().check(context.getExecution().getControl());
+      promise.onError(throwable -> {
+        hcheckResults.put(hcheck.get().getName(), HealthCheck.Result.unhealthy(throwable));
+        context.render(new HealthCheckResults(ImmutableSortedMap.<String, HealthCheck.Result>copyOfSorted(hcheckResults)));
```

Comment:
Can use Collections.singletonMap here instead of a concurrent map and then copying.

Same with the similar lines.